### PR TITLE
CSJ recipe to be compatible with lm

### PR DESCRIPTION
**Pull request by kan-bayashi**

### CODE REVIEW

File: egs/csj/asr1/run.sh [hunk 1]
```diff
@@ -151,8 +157,31 @@ if [ ${stage} -le 2 ]; then
...
+if [ ${stage} -le 3 ]; then
+    echo "stage 3: LM Preparation"
+    lmdatadir=data/local/lm_train
+    mkdir -p ${lmdatadir}
+    text2token.py -s 1 -n 1 data/${train_set}/text | cut -f 2- -d" " | perl -pe 's/\n/ <eos> /g' \
+        > ${lmdatadir}/train.txt
+    text2token.py -s 1 -n 1 data/${train_dev}/text | cut -f 2- -d" " | perl -pe 's/\n/ <eos> /g' \
+        > ${lmdatadir}/valid.txt
+    ${cuda_cmd} ${lmexpdir}/train.log \
+        lm_train.py \
```
Comment (sw005320) [inline egs/csj/asr1/run.sh]: It's better to play with batchsize (say 256 or 128). I put the batchsize extremely large for the WSJ task, but for CSJ, we can use smaller batchsize.

File: egs/csj/asr1/run.sh [hunk 2]
```diff
@@ -48,11 +50,15 @@ maxlen_out=150 # if output length > maxlen_out, batchsize is automatically reduc
 opt=adadelta
 epochs=15
 
+# rnnlm related
+lm_weight=0.3
```
Comment (sw005320) [inline egs/csj/asr1/run.sh]: is this value (0.3) obtained by testing several values?
If so, it is fine.
If not, you might test several values.

File: egs/csj/asr1/run.sh (hunk 1 above)
Comment (kan-bayashi) [reply egs/csj/asr1/run.sh]: OK. I will try small batchsize.

File: egs/csj/asr1/run.sh (hunk 2 above)
Comment (kan-bayashi) [reply egs/csj/asr1/run.sh]: Yes, I have tried {0.25, 0.3, 0.5, 1.0}, and 0.3 is the best.

**Comment by kan-bayashi** [comment]
I have tried lm training with small batch size.
The perplexity for validation data decreased compared with lm with large batch size, (16.65 -> 14.33) but the final CER is not so changed.  

**LM with large batch size**

 method | S | D | I | CER | S | D | I | CER | S| D | I | CER
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
decode_eval1_beam20_eacc.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm0.3/result.txt:\| | 5.3 | 2.2 | 1.3 | 8.8 | 4.1 | 1.2 | 0.9 | 6.2 | 4.2 | 1.5 | 1.2 | 7
decode_eval1_beam20_eacc.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm0.5/result.txt:\| | 5.3 | 2.1 | 1.4 | 8.9 | 4.1 | 1.3 | 1 | 6.4 | 4.4 | 1.6 | 1.4 | 7.4
decode_eval1_beam20_eacc.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm0.75/result.txt:\| | 5.6 | 2.1 | 1.8 | 9.6 | 4.4 | 1.4 | 1.2 | 6.9 | 4.8 | 1.6 | 1.7 | 8
decode_eval1_beam20_eacc.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm1.0/result.txt:\| | 6.3 | 2.1 | 2.3 | 10.7 | 5 | 1.4 | 1.5 | 7.9 | 5.4 | 1.7 | 2.1 | 9.2

**LM with small batch size**

method | S | D | I | CER | S | D | I | CER | S| D | I | CER
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
decode_eval1_beam20_eacc.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm0.3/result.txt:\| | 5.4 | 2.2 | 1.3 | 8.9 | 4 | 1.3 | 0.9 | 6.2 | 4.3 | 1.5 | 1.2 | 7
decode_eval1_beam20_eacc.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm0.5/result.txt:\| | 5.3 | 2.1 | 1.5 | 8.8 | 4 | 1.3 | 1 | 6.3 | 4.3 | 1.5 | 1.4 | 7.2
decode_eval1_beam20_eacc.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm0.75/result.txt:\| | 5.6 | 2 | 1.8 | 9.5 | 4.3 | 1.4 | 1.2 | 6.9 | 4.6 | 1.6 | 1.7 | 7.9
decode_eval1_beam20_eacc.best_p0.0_len0.0-0.0_ctcw0.3_rnnlm1.0/result.txt:\| | 6.2 | 2.1 | 2.4 | 10.7 | 5 | 1.4 | 1.7 | 8.2 | 5.2 | 1.7 | 2.1 | 9.1

File: egs/csj/asr1/run.sh [hunk 3]
```diff
@@ -151,8 +157,32 @@ if [ ${stage} -le 2 ]; then
          data/${train_dev} ${dict} > ${feat_dt_dir}/data.json
 fi
 
+# It takes a few days. If you just want to end-to-end ASR without LM,
```
Comment (sw005320) [inline egs/csj/asr1/run.sh]: I think you can remove this comment.
CSJ LM training would be finished quickly.

**Comment by kan-bayashi** [comment]
In the case of small batch, more epochs leads slightly better performance.
I updated results using this model.

**Pull request merged**